### PR TITLE
Support `date` validator

### PR DESCRIPTION
**Pull request by shimomuh**
I want to add date validator.

### CODE REVIEW

**Comment by r7kamura** [comment]
Hi @shimomuh! Long time no see, and it's great to see you again on GitHub 😃 

> I want to add date validator.

It's welcome to merge this useful feature, but the following ideas will bring more powers I think:

1. Supporting Time validator (instead of Date validator)
2. Supporting both a single format String and an Array of format Strings

```ruby
# Examples
time :a, only: "%Y-%m-%d"
time :b, only: "%Y-%m-%d %H:%M"
time :c, only: %w[%Y-%m-%d %Y/%m/%d]
```

@shimomuh How about these?

**Comment by shimomuh** [comment]
@r7kamura 
> lt's great to see you again on GitHub

me too 😄 

> the following ideas will bring more powers I think

Thank you for suggestion!
That's nice!

I want to distinguish between time and date.
```ruby
# Examples
date :a, only: '%Y-%m-%d'
time :b, only: '%Y-%m-%d %H:%M'
```

Accordingt to iso-8601 ( https://www.iso.org/iso-8601-date-and-time-format.html ), dates and times are distinguished as another word.
It does not matter if alias but I want to prepare each property!

@r7kamura How about you?

**Comment by r7kamura** [comment]
The more accurate distinction between date and time looks so good 👍 

In that case, `date :a, only: '%Y-%m-%d %H:%M'` is not appropriate code, so ideally we should not alias `date` to `time`...

File: README.md [hunk 1]
```diff
@@ -26,6 +26,8 @@ class RecipesController < ApplicationController
     string :quantity do |value|
       value =~ /\A\d+(?:\.\d+)g\z/
     end
+    date :date, only: ["%Y/%m/%d", "%Y.%m.%d"]
+    time :time, only: "%Y/%m/%d %H:%M:%S"
```
Comment (shimomuh) [inline README.md]: Support `String` and `Array`

File: lib/weak_parameters/time_validator.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module WeakParameters
+  class TimeValidator < WeakParameters::DateValidator
```
Comment (shimomuh) [inline lib/weak_parameters/time_validator.rb]: TimeValidator < DateValidator

File: lib/weak_parameters/date_validator.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    end
+
+    def date_formats
+      return %w[%Y-%m-%d] if options[:only].blank?
+      options[:only].is_a?(Array) ? options[:only] : [options[:only]]
+    end
+
+    def invalid_format?(date_format)
+      date_format =~ /(%H|%M|%S)/
+    end
```
Comment (shimomuh) [inline lib/weak_parameters/date_validator.rb]: If format includes hours, minutes or seconds, invalid as date (accurately)

**Comment by shimomuh** [comment]
> In that case, `date :a, only: '%Y-%m-%d %H:%M'` is not appropriate code, so ideally we should not alias `date` to `time` ...

I tried it!
@r7kamura  Please check and comment.

**Comment by r7kamura** [comment]
hmm, DEFALUT_FORMATS looks not to being used...

BTW, if any format is not given, we should use `Date.parse` and `Time.parse` instead like this:

```rb
module WeakParameters
  class DateValidator < WeakParameters::BaseValidator
    private

    def error_message
      "params[#{key.inspect}] must be a valid #{parser_class}"
    end

    # @return [Class]
    def parser_class
      ::Date
    end

    def valid_type?
      if options[:only]
        Array(options[:only]).any? do |format|
          begin
            parser_class.strptime(value, format)
            true
          rescue ::ArgumentError
            false
          end
        end
      else
        begin
          parser_class.parse(value)
          true
        rescue ::ArgumentError
          false
        end
      end
    end
  end
end

module WeakParameters
  class TimeValidator < WeakParameters::DateValidator
    private

    # @note Override
    def parser_class
      ::Time
    end
  end
end
```

**Comment by shimomuh** [comment]
> we should use `Date.parse` and `Time.parse`

It looks good and simple!

The code permits the following request:
```ruby
# no option
date: '2017-01-01 12:34:56' # valid (I shuold not permit)  
time: '2017-01-01' # valid (I can permit)

# with options[:only] = '%Y-%m-%d'
date: '2017-01-01 12:34:56' # valid (I should not permit)
```

cause:
```ruby
2.3.1 pry(main)> Date.parse('2017-01-01 12:34:56')
=> Sun, 01 Jan 2017
2.3.1 pry(main)> Time.parse('2017-01-01')
=> 2017-01-01 00:00:00 +0900

2.3.1 pry(main)> Date.strptime('2017-01-01 12:34:56', '%Y-%m-%d')
=> Sun, 01 Jan 2017
```

Can you permit these cases?
If cannot, I do following:
```ruby
module WeakParameters
  class DateValidator < WeakParameters::BaseValidator
    private

    def parser_class
      ::Date
    end

    def valid_type?
      if options[:only]
        Array(options[:only]).any? do |format|
          begin
            parser_class.strptime(value, format)
            return false unless strictly?
            true
          rescue ::ArgumentError
            false
          end
        end
      else
        begin
          parser_class.parse(value)
          return false unless strictly?
          true
        rescue ::ArgumentError
          false
        end
      end
    end

    def strictly?(format = nil)
      result = ::Date._strptime(value, format)
      result.present? && !result.key?(:leftover)
    end

    def error_message
      "params[#{key.inspect}] must be a valid Date"
    end
  end
end

module WeakParameters
  class TimeValidator < WeakParameters::DateValidator
    private

    # @note Override
    def parser_class
      ::Time
    end

    # time is always true!
    def strictly?
      true
    end
  end
end
```

@r7kamura How about this?
I want to decide your suggestion because simple.

**Comment by r7kamura** [comment]
Sorry for the delay in my response.
I agree the latter too, the more strict validation 👍

**Comment by shimomuh** [comment]
@r7kamura refixed!
Please check and comment!

File: lib/weak_parameters/date_validator.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+          true
+        rescue ::ArgumentError
+          false
+        end
+      end
+    end
+
+    def strictly?(format = nil)
+      result = format ? ::Date._strptime(value, format) : ::Date._strptime(value)
+      result.present? && !result.key?(:leftover)
```
Comment (r7kamura) [inline lib/weak_parameters/date_validator.rb]: We can't use `Hash#present?` here because ActiveSupport is not loaded in this library.

Comment (shimomuh) [reply lib/weak_parameters/date_validator.rb]: I'm so sorry.
refixed.

**Comment by shimomuh** [comment]
@r7kamura 
Thank you for your approval.
Please merge and release whenever you want!

**Comment by r7kamura** [comment]
👍 👍 👍

**Pull request merged**